Optim foreach cleanup for AdamW

mikaylagawarecki opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #70953
* __->__ #70484
* #70483
* #70482
* #70481
* #70295

Add foreach flag to AdamW optimizer + cleanup

Differential Revision: [D33767869](https://our.internmc.facebook.com/intern/diff/D33767869)

mikaylagawarecki committed:
Optim foreach cleanup for AdamW

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for AdamW"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for AdamW"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for AdamW"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for AdamW"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for AdamW"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for AdamW"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for AdamW"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for AdamW"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for AdamW"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for AdamW"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for AdamW"




[ghstack-poisoned]

albanD reviewed (approved):
SGTM

mikaylagawarecki commented on torch/optim/adamw.py:
```diff
@@ -141,18 +146,162 @@ def step(self, closure=None):
+        grads = torch._foreach_neg(tuple(grads))  # type: ignore[assignment]
+
+    # Perform stepweight decay
+    torch._foreach_mul_(params, 1 - lr * weight_decay)
```
I'm not sure whether this qualifies as a change, but initially this was  `p.mul_(1 - group['lr'] * group['weight_decay'])` for each p in params [here](https://github.com/pytorch/pytorch/blob/master/torch/optim/_multi_tensor/adamw.py#L83) but I made it foreach here since in the torch/optim/adamw class we perform stepweight decay within the functional form rather than the class

albanD replied:
Thanks for pointing it out.
That sounds ok!

ngimel replied:
Yeah that's a good change, no reason not to use foreach here

mikaylagawarecki committed:
Update on "Optim foreach cleanup for AdamW"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for AdamW"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for AdamW"


Differential Revision: [D33767869](https://our.internmc.facebook.com/intern/diff/D33767869)

[ghstack-poisoned]

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33767869).

mikaylagawarecki committed:
Update on "Optim foreach cleanup for AdamW"


Differential Revision: [D33767869](https://our.internmc.facebook.com/intern/diff/D33767869)

[ghstack-poisoned]

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33767869).
